### PR TITLE
[Core VM] update exception message in the vm to make it more clear.

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

Many of the exception messages in the VM lack details that can help user to locate the real problem, thus having this pr. 

## Type of change

- [ ] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo.VM/EvaluationStack.cs [hunk 1]
```diff
@@ -69,7 +69,7 @@ IEnumerator IEnumerable.GetEnumerator()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal void Insert(int index, StackItem item)
         {
-            if (index > innerList.Count) throw new InvalidOperationException($"Insert out of bounds: {index}/{innerList.Count}");
+            if (index > innerList.Count) throw new InvalidOperationException($"Insert index {index} is out of stack bounds: {index}/{innerList.Count}");
```
Comment (AnnaShaleva) [inline src/Neo.VM/EvaluationStack.cs:72]: This message contains duplicating information (`index`), I'd say it's excessive a bit.

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs [hunk 2]
```diff
@@ -46,19 +46,19 @@ public virtual void Memcpy(ExecutionEngine engine, Instruction instruction)
...
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The count {count} can not be negative for {nameof(OpCode.MEMCPY)}.");
             int si = (int)engine.Pop().GetInteger();
             if (si < 0)
-                throw new InvalidOperationException($"The value {si} is out of range.");
+                throw new InvalidOperationException($"The source index {si} can not be negative for {nameof(OpCode.MEMCPY)}.");
             ReadOnlySpan<byte> src = engine.Pop().GetSpan();
             if (checked(si + count) > src.Length)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The source index {si} + count {count} is out of range for {nameof(OpCode.MEMCPY)}, {si}/[0, {src.Length}).");
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Splice.cs:55]: > [0, {src.Length})

Right bound should be included, `]`.

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs [hunk 3]
```diff
@@ -46,19 +46,19 @@ public virtual void Memcpy(ExecutionEngine engine, Instruction instruction)
...
             int di = (int)engine.Pop().GetInteger();
             if (di < 0)
-                throw new InvalidOperationException($"The value {di} is out of range.");
+                throw new InvalidOperationException($"The destination index {di} can not be negative for {nameof(OpCode.MEMCPY)}.");
             Types.Buffer dst = engine.Pop<Types.Buffer>();
             if (checked(di + count) > dst.Size)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The destination index {di} + count {count} is out of range for {nameof(OpCode.MEMCPY)}, {di}/[0, {dst.Size}).");
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Splice.cs:61]: Ditto.

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs [hunk 4]
```diff
@@ -96,13 +96,13 @@ public virtual void SubStr(ExecutionEngine engine, Instruction instruction)
...
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The count {count} can not be negative for {nameof(OpCode.SUBSTR)}.");
             int index = (int)engine.Pop().GetInteger();
             if (index < 0)
-                throw new InvalidOperationException($"The value {index} is out of range.");
+                throw new InvalidOperationException($"The index {index} can not be negative for {nameof(OpCode.SUBSTR)}.");
             var x = engine.Pop().GetSpan();
             if (index + count > x.Length)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The index {index} + count {count} is out of range for {nameof(OpCode.SUBSTR)}, {index}/[0, {x.Length}).");
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Splice.cs:105]: Ditto.

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs [hunk 5]
```diff
@@ -120,10 +120,10 @@ public virtual void Left(ExecutionEngine engine, Instruction instruction)
         {
             int count = (int)engine.Pop().GetInteger();
             if (count < 0)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The count {count} can not be negative for {nameof(OpCode.LEFT)}.");
             var x = engine.Pop().GetSpan();
             if (count > x.Length)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The count {count} is out of range for {nameof(OpCode.LEFT)}, {count}/[0, {x.Length}).");
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Splice.cs:126]: Ditto.

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs [hunk 6]
```diff
@@ -141,10 +141,10 @@ public virtual void Right(ExecutionEngine engine, Instruction instruction)
         {
             int count = (int)engine.Pop().GetInteger();
             if (count < 0)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The count {count} can not be negative for {nameof(OpCode.RIGHT)}.");
             var x = engine.Pop().GetSpan();
             if (count > x.Length)
-                throw new InvalidOperationException($"The value {count} is out of range.");
+                throw new InvalidOperationException($"The count {count} is out of range for {nameof(OpCode.RIGHT)}, {count}/[0, {x.Length}).");
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Splice.cs:147]: Just check it everywhere.

File: src/Neo.VM/EvaluationStack.cs (hunk 1 above)
Comment (Jim8y) [reply src/Neo.VM/EvaluationStack.cs:72]: updated.

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs (hunk 3 above)
Comment (Jim8y) [reply src/Neo.VM/JumpTable/JumpTable.Splice.cs]: fixed

File: src/Neo.VM/JumpTable/JumpTable.Splice.cs (hunk 5 above)
Comment (Jim8y) [reply src/Neo.VM/JumpTable/JumpTable.Splice.cs]: fixed

File: src/Neo.VM/Types/InteropInterface.cs [hunk 7]
```diff
@@ -53,7 +53,7 @@ public override int GetHashCode()
         public override T GetInterface<T>()
         {
             if (_object is T t) return t;
-            throw new InvalidCastException($"The item can't be casted to type {typeof(T)}");
+            throw new InvalidCastException($"The item {_object} can't be casted to type {typeof(T)}.");
```
Comment (shargon) [inline src/Neo.VM/Types/InteropInterface.cs:56]: If _object.ToString fail, it change the exception

Comment (Jim8y) [reply src/Neo.VM/Types/InteropInterface.cs]: fixed.

**Review by shargon — APPROVED** [review]

**Review by nan01ab — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**